### PR TITLE
Build podutils with ko

### PR DESCRIPTION
**Pull request by chaodaiG**
Now that the git base image contains default github fingerprint, all pod utils images can be built with ko

/cc @cjwagner

### CODE REVIEW

**Comment by chaodaiG** [comment]
/retest
should be much stable after #25273 is merged

**Comment by chaodaiG** [comment]
Did a local testing, realized that the entrypoints of pod utils images are all overridden, for example:

- entrypoint: `cp /entrypoint /tools/entrypoint`
- clonerefs: `/clonerefs`
- initupload: `/initupload`
- sidecar: `/sidecar`

These are not the place where `ko` bin place images, which is `/ko-app`, so it won't work as is. Will need to figure out a way to ensure backward compatibility
/hold

**Comment by alvaroaleman** [comment]
I think a changed location will break some of the Openshift downstream tooling, cc @petr-muller 
What is the reason we excplicitly set the `command` rather than relying on `Entrypoint` being correct, that bazel doesn't set `Entrypoint`?

**Comment by chaodaiG** [comment]
> I think a changed location will break some of the Openshift downstream tooling, cc @petr-muller What is the reason we excplicitly set the `command` rather than relying on `Entrypoint` being correct, that bazel doesn't set `Entrypoint`?

I don't see a particular reason for setting `command` for `clonerefs`, `initupload`, and `sidecar`, just tested with `docker run -it gcr.io/k8s-prow/initupload` and the Entrypoint seemed to be correctly set, same with `clonerefs` and `sidecar`, I think they can safely be dropped. 

`entrypoint` image is different, the image does `cp /entrypoint /tools/entrypoint` instead of executing the binary, would the location change here breaking Openshift downstream tooling?

**Comment by alvaroaleman** [comment]
> entrypoint image is different, the image does cp /entrypoint /tools/entrypoint instead of executing the binary

A creative solution would be to add logic to `entrypoint` to copy itself to a location configured via a flag or env var?

I did some git archaeology and what broke last time the binary location changed is `clonerefs`, because we have code that copies it out of one image and copies it into another one. To avoid that in the future, there is now a downstream image that copies `clonerefs` to a well-known location, so that would need updating: https://github.com/openshift/release/blob/9a953c36206552da0d519981f95142b5e9447b07/clusters/app.ci/supplemental-ci-images/manage-clonerefs.yaml#L30

Other than that if we work around the problem by not setting the command, there is some downstream code that does decoration (because openshift has a testrunner that creates more pods that are also decorated), but _I think_ that should just work once it has an updated t-i dependency: https://github.com/openshift/ci-tools/blob/e5be54a1382f18b156ed3f41ae2808bb9e4770d7/pkg/steps/artifacts.go#L395

**Comment by chaodaiG** [comment]
> > entrypoint image is different, the image does cp /entrypoint /tools/entrypoint instead of executing the binary
> 
> A creative solution would be to add logic to `entrypoint` to copy itself to a location configured via a flag or env var?

I have thought about this as well :). This would for sure work I believe, and the problem is that it'll be breaking change, the command of entrypoint image will be changed from:
```
- command: cp
  arg:
  - /entrypoint
  - /tools/entrypoint
```
to

```
- arg:
  - --copy-myself=/tools/entrypoint
```
Even if we make it smart enough to operate in these two modes, the old pod utils images will not work with new decorating logic. Any idea how to solve this


> I did some git archaeology and what broke last time the binary location changed is `clonerefs`, because we have code that copies it out of one image and copies it into another one. To avoid that in the future, there is now a downstream image that copies `clonerefs` to a well-known location, so that would need updating: https://github.com/openshift/release/blob/9a953c36206552da0d519981f95142b5e9447b07/clusters/app.ci/supplemental-ci-images/manage-clonerefs.yaml#L30
> 
> Other than that if we work around the problem by not setting the command, there is some downstream code that does decoration (because openshift has a testrunner that creates more pods that are also decorated), but _I think_ that should just work once it has an updated t-i dependency: https://github.com/openshift/ci-tools/blob/e5be54a1382f18b156ed3f41ae2808bb9e4770d7/pkg/steps/artifacts.go#L395

Thank you for the context

**Comment by alvaroaleman** [comment]
> Even if we make it smart enough to operate in these two modes, the old pod utils images will not work with new decorating logic.

That is true, but do we assume there are Prow setups that update prow itself but do not update the decoration utils images at the same time? Because that is the only situation where this would be a problem or not?

**Comment by chaodaiG** [comment]
> > Even if we make it smart enough to operate in these two modes, the old pod utils images will not work with new decorating logic.
> 
> That is true, but do we assume there are Prow setups that update prow itself but do not update the decoration utils images at the same time? Because that is the only situation where this would be a problem or not?

Yeah technically we don't necessarily need to support mismatched versions of prow component and decoration components. There is a slightly chance of race, that prow components and decoration configs are not guaranteed to be updated at the same time, so the jobs scheduled in between would fail, but on the other hand this window should be very narrow. @cjwagner , wdyt?

**Comment by cjwagner** [comment]
> Yeah technically we don't necessarily need to support mismatched versions of prow component and decoration components.

We shouldn't worry about supporting this at all in my opinion. I think the short window of version mismatch is fine. Now that we use `Make` to deploy, the deployment job is much faster. We could even do a manual deploy right when the PR merges if we want to reduce downtime further.

**Comment by chaodaiG** [comment]
Cool, see https://github.com/kubernetes/test-infra/pull/25305

**Comment by chaodaiG** [comment]
The blockers were removed by #25305  and #25301 , this PR is now pretty simple
/unhold

**Comment by alvaroaleman** [comment]
/hold
downstream consumers of podutils need to be updated before we merge this

File: hack/prowimagebuilder/main.go
Comment (BenTheElder) [inline hack/prowimagebuilder/main.go:52]: FWIW it is useful even when not pushing to have a predictable and reasonable name so tarballs can be `docker load` ed for development testing (or `kind load image-archive`).

Comment (BenTheElder) [reply hack/prowimagebuilder/main.go:52]: I typically use the location I _would_ have pushed to.

Comment (chaodaiG) [reply hack/prowimagebuilder/main.go:52]: For some reason ko failed to build multi-arch images into tarballs, so I chose this as alternative, hopefully no one needs to `docker load` tarball for development

**Comment by petr-muller** [comment]
Thanks for holding this until we can adapt. Would it be a problem if this waited for next week? Most of RH engineering will not be working tomorrow and this looks like can cause a pretty stressful outage if we overlook something.

**Comment by chaodaiG** [comment]
> Thanks for holding this until we can adapt. Would it be a problem if this waited for next week? Most of RH engineering will not be working tomorrow and this looks like can cause a pretty stressful outage if we overlook something.

I think it's fine to wait until next week. The last thing I want to see is breaking downstream and had to revert upstream, repeatedly. Thank you @alvaroaleman and @petr-muller  for thinking ahead of time!

**Comment by petr-muller** [comment]
OK, we're prepared on our side. Thanks for holding this!

**Comment by chaodaiG** [comment]
> OK, we're prepared on our side. Thanks for holding this!

Awesome!
/unhold

**Comment by chaodaiG** [comment]
@cjwagner @alvaroaleman , could you review?

**Review by alvaroaleman — APPROVED** [review]

**Pull request merged**
